### PR TITLE
Enum discriminant

### PR DESCRIPTION
**Pull request by essickmango**
Implements the MiniRust interface for enum discriminants, namely get_discriminant (ValueExpr::Discriminant) and set_discriminant (Statement::SetDiscriminant).

Adds tests for the discriminant interface.

### CODE REVIEW

File: tooling/minitest/src/tests/enum_discriminant.rs [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+            assign(downcast(field(local(0), 0), 0), const_int(42u8)),
+            set_discriminant(field(local(0), 0), 0),
+            if_(eq(load(field(local(0), 1)), const_int(42u8)), 2, 3),
+        ),
+        block!(exit()),
+        block!(unreachable()),
+    ];
+    let program = program(&[function(Ret::No, 0, &locals, &blocks)]);
+    assert_stop(program);
+}
```
Comment (RalfJung) [inline tooling/minitest/src/tests/enum_discriminant.rs:214]: Missing newline at end of file

File: tooling/miniutil/src/build/ty.rs [hunk 2]
```diff
@@ -70,10 +70,11 @@ pub fn enum_variant(ty: Type, tagger: &[(Offset, u8)]) -> Variant {
     }
 }
 
-pub fn enum_ty(variants: &[Variant], discriminator: Discriminator, size: Size, align: Align) -> Type {
+pub fn enum_ty(variants: &[Variant], discriminator: Discriminator, discriminant_bytes: impl Into<Int>, size: Size, align: Align) -> Type {
```
Comment (RalfJung) [inline tooling/miniutil/src/build/ty.rs:73]: Wouldn't it be better to give the discriminant type as a type? Callers would then do something like `<u8>::get_type()`. Or maybe make `enum_ty` take a `T: TypeConv` then one would do `enum_ty::<u8>(...)`.

File: spec/lang/step/expressions.md [hunk 3]
```diff
@@ -94,6 +94,34 @@ impl<M: Memory> Machine<M> {
...
+        let (place, ty) = self.eval_place(place)?;
+        let Type::Enum { discriminator, discriminant_ty, .. } = ty else {
+            panic!();
+        };
+
+        // We don't require the variant to be valid,
+        // we are only interested in the bytes that the discriminator actually touches.
+        let accessor = |idx| {
+            let ptr = self.ptr_offset_inbounds(place.ptr, idx)?;
+            let byte = self.mem.load(ptr, Size::from_bytes(1).unwrap(), Align::ONE, Atomicity::None)?;
```
[the remark this answers is not in the document]
Comment (essickmango) [reply spec/lang/step/expressions.md:113]: I have not found those alignment requirements other than that the tag has to be aligned according to its representation? Please give me a hint where I can find them if I was blind, they don't seem to be anywhere in the lowering or computation of the tag layout/representation.

However I don't see how these alignment requirements play a role when we read each byte for the decision tree individually. They would have to be ensured earlier, where we know more about the generation of the decision tree, not?

[the remark this answers is not in the document]
Comment (RalfJung) [reply spec/lang/step/expressions.md:113]: Don't worry about alignment for this PR, just add the FIXME so we remember to get back to it later. We can discuss this in our next meeting.

[the remark this answers is not in the document]
Comment (essickmango) [reply spec/lang/step/expressions.md:113]: okay

File: tooling/miniutil/src/build/ty.rs (hunk 2 above)
Comment (essickmango) [reply tooling/miniutil/src/build/ty.rs:73]: I like the idea of making `enum_ty` take a `DiscriminantTy: TypeConv` as this type is part of the enum. (I also realized that the current way doesn't allow signed integer discriminants without the bare-bones construction.)

**Comment by RalfJung** [comment]
Looks good, thanks. :)

**Pull request merged**